src/main/java/com/conveyal/gtfs/loader/JdbcGtfsSnapshotter.java

```diff
@@ -274,42 +276,52 @@ private TableLoadResult createScheduleExceptionsTable() {
                 }
                 scheduleExceptionsTracker.executeRemaining();
 
-                // determine if we appear to be working with a calendar_dates-only feed.
-                // If so, we must also add dummy entries to the calendar table
-                if (
-                    feedIdToSnapshot != null &&
-                    !tableExists(feedIdToSnapshot, "calendar") &&
-                    calendarDatesReader.getRowCount() > 0
-                ) {
-                    sql = String.format(
-                        "insert into %s (service_id, description, start_date, end_date, " +
-                            "monday, tuesday, wednesday, thursday, friday, saturday, sunday)" +
-                            "values (?, ?, ?, ?, 0, 0, 0, 0, 0, 0, 0)",
-                        tablePrefix + "calendar"
+                // fetch all entries in the calendar table to generate set of serviceIds that exist in the calendar
+                // table.
+                JDBCTableReader<Calendar> calendarReader = new JDBCTableReader(
+                    Table.CALENDAR,
+                    dataSource,
+                    feedIdToSnapshot + ".",
+                    EntityPopulator.CALENDAR
+                );
+                Set<String> calendarServiceIds = new HashSet<>();
+                for (Calendar calendar : calendarReader.getAll()) {
+                    calendarServiceIds.add(calendar.service_id);
+                }
+
+                // add auto-generated entries to the calendar table that only existed in the calendar_dates table
+                sql = String.format(
+                    "insert into %s (service_id, description, start_date, end_date, " +
+                        "monday, tuesday, wednesday, thursday, friday, saturday, sunday)" +
+                        "values (?, ?, ?, ?, 0, 0, 0, 0, 0, 0, 0)",
+                    tablePrefix + "calendar"
+                );
+                PreparedStatement calendarStatement = connection.prepareStatement(sql);
+                final BatchTracker calendarsTracker = new BatchTracker(
+                    "calendar",
+                    calendarStatement
+                );
+                for (Calendar calendar : calendarsByServiceId.values()) {
```

Comment:
Could we change the variable name `calendarsByServiceId` to `dummyCalendarsByServiceId`? I had to track up in this PR to remind myself what this map contained.